devices/cpu/registers.c

```diff
@@ -200,3 +204,101 @@ void set_sr_from_fields(Emulator *emu, const Status_reg fields)
 
   cpu->sr = r2;
 }
+
+void update_cpu_stats(Emulator* const emu)
+{
+  char buffer[STRING_BUFFER_SIZE];
+  if (emu->cpu->stats.spLowWatermark > emu->cpu->sp)
+  {
+    if (emu->do_trace)
+    {
+      sprintf(buffer, "New SP low watermark - %04X\n", emu->cpu->sp);
+      print_console(emu, buffer);
+    }
+    emu->cpu->stats.spLowWatermark = emu->cpu->sp;
+  }
+
+  emu->cpu->stats.spLastValue = emu->cpu->sp;
+}
+
+void display_cpu_stats(Emulator* const emu)
+{
+  char stats[STRING_BUFFER_SIZE];
+  sprintf(stats, "CPU stats:\n \tSP low watermark - %04X\n",
+    emu->cpu->stats.spLowWatermark);
+  print_console(emu, stats);
+}
+
+void reset_cpu_stats(Emulator* const emu)
+{
+  emu->cpu->stats.spLowWatermark = 0xFFFF;
+  emu->cpu->stats.spLastValue = 0xFFFF;
+}
+
+void reset_call_tracer(Emulator* const emu)
+{
+  emu->cpu->callTracer.callDepth = 0;
+}
+
+static void print_spaces(Emulator* const emu, const uint8_t count)
+{
+  for (uint8_t i = 0; i < count; i++)
+    print_console(emu, " ");
+}
+
+static void reportCall(Emulator* const emu)
+{
+  char buffer[STRING_BUFFER_SIZE];
+  Cpu* const cpu = emu->cpu;
+  print_spaces(emu, cpu->callTracer.callDepth);
+  sprintf(buffer, "Call from %04X to %04X; [SP] == %04X\n",
+    cpu->callTracer.calls[cpu->callTracer.callDepth].returnPc,
+    cpu->callTracer.calls[cpu->callTracer.callDepth].targetPc,
+    cpu->callTracer.calls[cpu->callTracer.callDepth].sp);
+  print_console(emu, buffer);
+}
+
+static void reportReturn(Emulator* const emu)
+{
+  char buffer[STRING_BUFFER_SIZE];
+  Cpu* const cpu = emu->cpu;
```

Comment:
`const Cpu`? it's used for read-only (same in function above)

Reply:
ok